### PR TITLE
Delete @storybook/react

### PR DESCRIPTION
**Pull request by MichaelDeBoey**
Please fill in this template.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).

Select one of these and delete the others:

If removing a declaration:
- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

---
Since #38961 is closed automatically, this one supersedes that one.

> As a follow-up of #38447, I now remove `@storybook/react` in a separate PR, where we can have an isolated look at how we could fix the errors.

CC @gaetanmaisse @kroeder @ndelangen @shilman @emilio-martinez @lonyele @benoitdion @dandean @sairus2k @ExE-Boss

### CODE REVIEW

**Comment by gaetanmaisse** [comment]
@MichaelDeBoey https://github.com/DefinitelyTyped/DefinitelyTyped/pull/39310 was merged 🎉 you can update your branch!

File: types/storybook-react-router/index.d.ts [hunk 1]
```diff
@@ -2,18 +2,18 @@
...
-export const StoryRouter: React.ComponentType<{
-  story: Story;
+export const StoryRouter: ComponentType<{
+  story: StoryApi;
   links: object;
   routerProps: MemoryRouterProps;
 }>;
 
-declare function storyRouterDecorator(links?: object, routerProps?: MemoryRouterProps): StoryDecorator;
+declare function storyRouterDecorator(links?: object, routerProps?: MemoryRouterProps): DecoratorFunction;
```
Comment (gaetanmaisse) [inline types/storybook-react-router/index.d.ts]: `DecoratorFunction` is a generic class and here we should use `DecoratorFunction<StoryFnReactReturnType>`. However `StoryFnReactReturnType` is not exported from `@storybook/react` so I think we can use `DecoratorFunction<any>` for now.

Comment (ndelangen) [reply types/storybook-react-router/index.d.ts]: StoryFnReactReturnType = ReactElement<unknown>

File: types/storybook-readme/index.d.ts [hunk 2]
```diff
@@ -2,10 +2,12 @@
...
 // Definitions by: Taeheon Kim  <https://github.com/lonyele>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.8
+// TypeScript Version: 3.1
+
+import { ComponentType, ReactNode } from 'react';
 
 // Shared Types
-export type Renderable = React.ComponentType | JSX.Element;
+export type Renderable = ComponentType | JSX.Element;
```
Comment (gaetanmaisse) [inline types/storybook-readme/index.d.ts]: For `storybook-readme` related issues (https://travis-ci.org/DefinitelyTyped/DefinitelyTyped/builds/604709995#L588) I'm sure we can ask @lonyele some help to update this typings to make them work with SB own types ;)

File: types/storybook-react-router/index.d.ts (hunk 1 above)
Comment (MichaelDeBoey) [reply types/storybook-react-router/index.d.ts]: I can import it from `'@storybook/react/dist/client/preview/types'`

```ts
import { StoryFnReactReturnType } from '@storybook/react/dist/client/preview/types';
```

**Comment by gaetanmaisse** [comment]
@MichaelDeBoey  It looks like `esModuleInterop` flag is missing in tsconfig of `storybook__addon-storyshots-puppeteer`

**Comment by MichaelDeBoey** [comment]
@gaetanmaisse @ndelangen Any idea how I can get this one fixed? 🤔

**Comment by gaetanmaisse** [comment]
@MichaelDeBoey I will take a look this weekend ;)

**Comment by MichaelDeBoey** [comment]
That would be awesome @gaetanmaisse! 👊

**Comment by gaetanmaisse** [comment]
Just have to wait for https://github.com/DefinitelyTyped/DefinitelyTyped/pull/40270 to be merged and we surely will be good to go!

(sadly it looks like `dtslint` used in CI is broken for scoped package 😞, I put a message in [Gitter](https://gitter.im/DefinitelyTyped/DefinitelyTyped), hope it will be back soon)

**Pull request closed without merging**